### PR TITLE
conform hostname/nodename to DNS-1123 subdomain spec

### LINKED ISSUE #57765: OpenStack cloud provider should register with instance id instead of instance name
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

The OpenStack cloud provider currently uses the name of the kubelet node to register with the API server. However the name is not guaranteed to conform to the DNS-1123 subdomain spec.

Given a node name of `worker/925f11f3-80af-4844-a00e-f0af521779d1`, the following message appears in the kubelet logs:

```
Unable to register node "worker/925f11f3-80af-4844-a00e-f0af521779d1" with API server: Node "worker/925f11f3-80af-4844-a00e-f0af521779d1" is invalid: metadata.name: Invalid value: "worker/925f11f3-80af-4844-a00e-f0af521779d1": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is 'a-z0-9?(.a-z0-9?)*')
```

**What you expected to happen**:

The OpenStack Cloud provider uses the instance id instead of the instance name. This will guarantee that the id conforms the DNS-1123 spec without making any assumptions on what the instance name may be.

**How to reproduce it (as minimally and precisely as possible)**:

In the OpenStack console, rename a kubelet node to `kubelet/1234` and attempt to register the kubelet node to the api server.

**Anything else we need to know?**:

**Environment**: OpenStack Newton
- Kubernetes version (use `kubectl version`): 

```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: OpenStack
- OS (e.g. from /etc/os-release): Ubuntu 14.04.5
- Kernel (e.g. `uname -a`): Linux b4fdd4bc-cab9-44ff-9c99-44e91ff5b414 4.4.0-103-generic #126~14.04.1-Ubuntu SMP Mon Dec 4 19:33:04 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
Currently the `hostName` can be arbitrary strings, even with `--hostnameOverride`, which should conform DNS-1123 subdomain spec.

This should also apply to `nodeName` for cloud-based nodes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #57765

**Special notes for your reviewer**:
/cc @smarterclayton @lavalamp 

**Release note**:

```release-note
conform hostname/nodename to DNS-1123 subdomain spec
```

### CODE REVIEW

File: pkg/util/node/node.go [hunk 1]
```diff
@@ -52,7 +54,36 @@ func GetHostname(hostnameOverride string) string {
...
+	nodeName, err := instances.CurrentNodeName(hostname)
+	if err != nil {
+		return "", fmt.Errorf("error fetching current node name from cloud provider: %v", err)
+	}
+
+	if msgs := utilvalidation.IsDNS1123Subdomain(string(nodeName)); len(msgs) != 0 {
+		return "", fmt.Errorf("invalid nodename: %s", strings.Join(msgs, ", "))
+	}
+
+	glog.V(2).Infof("cloud provider determined current node name to be %s", nodeName)
```
Comment (xiangpengzhao) [inline pkg/util/node/node.go]: The first letter of log info should be upper-case. (the original code is lower-case though)

**Review by xiangpengzhao — COMMENTED** [review]
just a nit, otherwise LGTM. Defer to other folks to take another look.

**Comment by dixudx** [comment]
/assign @dims 

Ping  @FengyunPan @smarterclayton for review. Thanks.

**Comment by dixudx** [comment]
Ping @FengyunPan @smarterclayton for review. Thanks.

**Pull request closed without merging**